mypy/plugin.py

```diff
@@ -379,3 +421,332 @@ def int_pow_callback(ctx: MethodContext) -> Type:
+    'attr.attrib',
+    'attr.attr',
+}
+
+
+class Attribute:
+    """The value of an attr.ib() call."""
+
+    def __init__(self, name: str, type: Optional[Type],
+                 has_default: bool, init: bool,
+                 context: Context) -> None:
+        self.name = name
+        self.type = type
+        self.has_default = has_default
+        self.init = init
+        self.context = context
+
+    def argument(self) -> Argument:
+        """Return this attribute as an argument to __init__."""
+        # Convert type not set to Any.
+        _type = self.type or AnyType(TypeOfAny.unannotated)
+        # Attrs removes leading underscores when creating the __init__ arguments.
+        return Argument(Var(self.name.strip("_"), _type), _type,
+                        None,
+                        ARG_OPT if self.has_default else ARG_POS)
+
+
+def attr_class_maker_callback(
+        attr_classes: Dict[TypeInfo, List[Attribute]],
+        ctx: ClassDefContext,
+        auto_attribs_default: bool = False
+) -> None:
+    """Add necessary dunder methods to classes decorated with attr.s.
+
+    Currently supports init=, cmp= and frozen= and auto_attribs=.
+    """
+    # attrs is a package that lets you define classes without writing dull boilerplate code.
+    #
+    # At a quick glance, the decorator searches the class body for assignments of `attr.ib`s (or
+    # annotated variables if auto_attribs=True), then depending on how the decorator is called,
+    # it will add an __init__ or all the __cmp__ methods.  For frozen=True it will turn the attrs
+    # into properties.
+    #
+    # See http://www.attrs.org/en/stable/how-does-it-work.html for information on how attrs works.
+
+    def called_function(expr: Expression) -> Optional[str]:
+        """Return the full name of the function being called by the expr, or None."""
+        if isinstance(expr, CallExpr) and isinstance(expr.callee, RefExpr):
+            return expr.callee.fullname
+        return None
+
+    def get_callable_type(call: CallExpr) -> Optional[CallableType]:
+        """Get the CallableType that's attached to a CallExpr."""
+        callable_type = None
+        if (isinstance(call, CallExpr)
+                and isinstance(call.callee, RefExpr)
+                and isinstance(call.callee.node, Var)
+                and call.callee.node.type):
+            callee_type = call.callee.node.type
+            if isinstance(callee_type, Overloaded):
+                # We take the last overload.
+                callable_type = callee_type.items()[-1]
+            elif isinstance(callee_type, CallableType):
+                callable_type = callee_type
+        return callable_type
+
+    def get_argument(call: CallExpr, name: str) -> Optional[Expression]:
+        """Return the expression for the specific argument."""
+        callee = get_callable_type(call)
+        if not callee:
+            return None
+
+        argument = callee.argument_by_name(name)
```

Comment:
I would try to inline this function instead of using ``get_callable_type`` (unless it is too long). I think the corresponding ``FuncDef`` should have all the information

Reply:
Which `FuncDef` are you referring to?

Reply:
The one `callee` normally "points to". I think you see some `Var`s because you use function aliases, but even in that case you should be able to get to the original `FuncDef`.

Reply:
So 2 problems with this approach.  

1. In the non-`Var` case, `call.callee.node` points to an `OverloadedFuncDef`.  So I still need to pick one of the overloads to get a `FuncDef` is that an ok thing to do?

2. For the life of me the only way I can see to get a `FuncDef` from a `Var` is to use `lookup_qualified` to get the `SymbolTableNode` get the `CallableType` out of that and then get the `FuncDef` from `definition` Is that what you had in mind?  I can also instead get rid of the assignments in the stubs and just repeat the function definitions for all of them.  (That's the way the real stubs are written anyway because of a bug in pycharm) But I'd rather not depend on that.

Another option is to forget the whole looking at the stub file and just hard code the positions.  This will mean this can't be used as is for future extensions but why code for something that hasn't been developed.

Let me know what you think.

Reply:
My intuition here is that if there will be a major change in `attrs`, then this plugin will anyway needs to be updated. So hard-coding some function signatures is OK. Also IIUC this is the logic other smaller plugins use (like `@contextlib.contextmanager`).

But I don't want to push you on this, if you believe it is better to extract signatures from the stubs then it is also OK, if the code readability doesn't suffer too much.